packages/web/src/components/apps/use-logs.tsx

```diff
@@ -0,0 +1,75 @@
+import React, { createContext, useContext, useEffect, useState } from 'react';
+
+import { AppChannel } from '@/clients/websocket';
+import { PreviewStatusPayloadType } from '@srcbook/shared';
+
+export type LogMessage = {
+  type: 'npm_install_error' | 'vite_error'; // TODO: add more types like "warning" or "problem"
+  timestamp: Date;
+  contents: string;
+};
+
+export interface LogsContextValue {
+  logs: Array<LogMessage>;
+  clearLogs: () => void;
+  addError: (message: Omit<LogMessage, 'timestamp'>) => void;
+  unreadLogsCount: number;
+
+  open: boolean;
+  togglePane: () => void;
+}
+
+const LogsContext = createContext<LogsContextValue | undefined>(undefined);
+
+type ProviderPropsType = {
+  channel: AppChannel;
+  children: React.ReactNode;
+};
+
+export function LogsProvider({ channel, children }: ProviderPropsType) {
+  const [logs, setLogs] = useState<Array<LogMessage>>([]);
+  const [unreadLogsCount, setUnreadLogsCount] = useState(0);
+
+  const [open, setOpen] = useState(false);
+
+  function clearLogs() {
+    setLogs([]);
+    setUnreadLogsCount(0);
+  }
+
+  function addError(error: Omit<LogMessage, 'timestamp'>) {
+    setLogs((logs) => [{ ...error, timestamp: new Date() }, ...logs]);
+    setUnreadLogsCount((n) => n + 1);
+  }
+
+  function togglePane() {
+    setOpen((n) => !n);
+    setUnreadLogsCount(0);
+  }
+
+  // If vite crashes, then create an error log
+  useEffect(() => {
```

Comment:
Fine for now (I don't have a better suggestion), but somehow feels like a weird place to have this listener